In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression

# Pandas library is used for importing various file types e.g. csv, excel, json.
# also used for manipulation and cleaning dataframes.

# Numpy library has functions used for working with arrays and matrices
# faster, clearer and better quality code using Numpy

# Matplotlib is a 2D plotting library for visualising data in Python

# Statsmodels.api provides classes and functions for the estimation of many different statistical models,
# as well as for conducting statistical tests, and statistical data exploration

# Seaborn is used for improving Matplotlib visualisations to make them more visually attractive.
# It builds on top of MatPlotLib and works with Pandas for data structures

# sklearn is a library used for machine learning and statistical modeling.
# it includes many tools for classification, regression, clustering and dimensionality reduction

In [2]:
data = pd.read_csv("Life Expectancy Data .csv")
data

# importing and displaying the data

Country  Year      Status  LifeExpectancy  AdultMortality  \
0     Afghanistan  2015  Developing            65.0           263.0   
1     Afghanistan  2014  Developing            59.9           271.0   
2     Afghanistan  2013  Developing            59.9           268.0   
3     Afghanistan  2012  Developing            59.5           272.0   
4     Afghanistan  2011  Developing            59.2           275.0   
...           ...   ...         ...             ...             ...   
2933     Zimbabwe  2004  Developing            44.3           723.0   
2934     Zimbabwe  2003  Developing            44.5           715.0   
2935     Zimbabwe  2002  Developing            44.8            73.0   
2936     Zimbabwe  2001  Developing            45.3           686.0   
2937     Zimbabwe  2000  Developing            46.0           665.0   

      infantDeaths  Alcohol  PercentageExpenditure  HepatitisB  Measles  ...  \
0               62     0.01              71.279624        65.0     1154  ...   
1               64     0.01              73.523582        62.0      492  ...   
2               66     0.01              73.219243        64.0      430  ...   
3               69     0.01              78.184215        67.0     2787  ...   
4               71     0.01               7.097109        68.0     3013  ...   
...            ...      ...                    ...         ...      ...  ...   
2933            27     4.36               0.000000        68.0       31  ...   
2934            26     4.06               0.000000         7.0      998  ...   
2935            25     4.43               0.000000        73.0      304  ...   
2936            25     1.72               0.000000        76.0      529  ...   
2937            24     1.68               0.000000        79.0     1483  ...   

      Polio  TotalExpenditure  Diphtheria   HIV         GDP  Population  \
0       6.0              8.16        65.0   0.1  584.259210  33736494.0   
1      58.0              8.18        62.0   0.1  612.696514    327582.0   
2      62.0              8.13        64.0   0.1  631.744976  31731688.0   
3      67.0              8.52        67.0   0.1  669.959000   3696958.0   
4      68.0              7.87        68.0   0.1   63.537231   2978599.0   
...     ...               ...         ...   ...         ...         ...   
2933   67.0              7.13        65.0  33.6  454.366654  12777511.0   
2934    7.0              6.52        68.0  36.7  453.351155  12633897.0   
2935   73.0              6.53        71.0  39.8   57.348340    125525.0   
2936   76.0              6.16        75.0  42.1  548.587312  12366165.0   
2937   78.0              7.10        78.0  43.5  547.358878  12222251.0   

      Thinness1to19Years  Thinness5to9Years  IncomeCompositionOfResources  \
0                   17.2               17.3                         0.479   
1                   17.5               17.5                         0.476   
2                   17.7               17.7                         0.470   
3                   17.9               18.0                         0.463   
4                   18.2               18.2                         0.454   
...                  ...                ...                           ...   
2933                 9.4                9.4                         0.407   
2934                 9.8                9.9                         0.418   
2935                 1.2                1.3                         0.427   
2936                 1.6                1.7                         0.427   
2937                11.0               11.2                         0.434   

      Schooling  
0          10.1  
1          10.0  
2           9.9  
3           9.8  
4           9.5  
...         ...  
2933        9.2  
2934        9.5  
2935       10.0  
2936        9.8  
2937        9.8  

[2938 rows x 22 columns]

### Cleaning

In [3]:
data.isnull().sum().sum()

# counts the total number of null values in dataset

2563

#### Update: need to convert non-numeric values to numeric in order to clean

In [4]:
# Identify non-numeric values
non_numeric_values = data['Country'].loc[~pd.to_numeric(data['Country'], errors='coerce').notnull()]

# Display non-numeric values
print("Non-numeric values:", non_numeric_values)

# Handle missing or invalid values
data['Country'] = data['Country'].replace(['N/A', 'NaN', 'Invalid'], pd.NA)  # Replace with Pandas NA

# Convert data to numeric type
data['Country'] = pd.to_numeric(data['Country'], errors='coerce')

# Now you can perform numeric operations on the column without encountering errors

Non-numeric values: 0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
2933       Zimbabwe
2934       Zimbabwe
2935       Zimbabwe
2936       Zimbabwe
2937       Zimbabwe
Name: Country, Length: 2938, dtype: object


In [5]:
# Identify non-numeric values
non_numeric_values = data['Status'].loc[~pd.to_numeric(data['Status'], errors='coerce').notnull()]

# Display non-numeric values
print("Non-numeric values:", non_numeric_values)

# Handle missing or invalid values
data['Status'] = data['Status'].replace(['N/A', 'NaN', 'Invalid'], pd.NA)  # Replace with Pandas NA

# Convert data to numeric type
data['Status'] = pd.to_numeric(data['Status'], errors='coerce')

# Now you can perform numeric operations on the column without encountering errors

Non-numeric values: 0       Developing
1       Developing
2       Developing
3       Developing
4       Developing
           ...    
2933    Developing
2934    Developing
2935    Developing
2936    Developing
2937    Developing
Name: Status, Length: 2938, dtype: object


In [6]:
data.fillna(data.mean(), inplace=True)

# counts the number of null values in dataset by feature

In [7]:
data.fillna(data.mean())

# since there are many missing values (nulls), the mean of each collumn will be used to fill in the missing values
# (this may not be the best way to fill the missing values)

Country  Year  Status  LifeExpectancy  AdultMortality  infantDeaths  \
0         NaN  2015     NaN            65.0           263.0            62   
1         NaN  2014     NaN            59.9           271.0            64   
2         NaN  2013     NaN            59.9           268.0            66   
3         NaN  2012     NaN            59.5           272.0            69   
4         NaN  2011     NaN            59.2           275.0            71   
...       ...   ...     ...             ...             ...           ...   
2933      NaN  2004     NaN            44.3           723.0            27   
2934      NaN  2003     NaN            44.5           715.0            26   
2935      NaN  2002     NaN            44.8            73.0            25   
2936      NaN  2001     NaN            45.3           686.0            25   
2937      NaN  2000     NaN            46.0           665.0            24   

      Alcohol  PercentageExpenditure  HepatitisB  Measles  ...  Polio  \
0        0.01              71.279624        65.0     1154  ...    6.0   
1        0.01              73.523582        62.0      492  ...   58.0   
2        0.01              73.219243        64.0      430  ...   62.0   
3        0.01              78.184215        67.0     2787  ...   67.0   
4        0.01               7.097109        68.0     3013  ...   68.0   
...       ...                    ...         ...      ...  ...    ...   
2933     4.36               0.000000        68.0       31  ...   67.0   
2934     4.06               0.000000         7.0      998  ...    7.0   
2935     4.43               0.000000        73.0      304  ...   73.0   
2936     1.72               0.000000        76.0      529  ...   76.0   
2937     1.68               0.000000        79.0     1483  ...   78.0   

      TotalExpenditure  Diphtheria   HIV         GDP  Population  \
0                 8.16        65.0   0.1  584.259210  33736494.0   
1                 8.18        62.0   0.1  612.696514    327582.0   
2                 8.13        64.0   0.1  631.744976  31731688.0   
3                 8.52        67.0   0.1  669.959000   3696958.0   
4                 7.87        68.0   0.1   63.537231   2978599.0   
...                ...         ...   ...         ...         ...   
2933              7.13        65.0  33.6  454.366654  12777511.0   
2934              6.52        68.0  36.7  453.351155  12633897.0   
2935              6.53        71.0  39.8   57.348340    125525.0   
2936              6.16        75.0  42.1  548.587312  12366165.0   
2937              7.10        78.0  43.5  547.358878  12222251.0   

      Thinness1to19Years  Thinness5to9Years  IncomeCompositionOfResources  \
0                   17.2               17.3                         0.479   
1                   17.5               17.5                         0.476   
2                   17.7               17.7                         0.470   
3                   17.9               18.0                         0.463   
4                   18.2               18.2                         0.454   
...                  ...                ...                           ...   
2933                 9.4                9.4                         0.407   
2934                 9.8                9.9                         0.418   
2935                 1.2                1.3                         0.427   
2936                 1.6                1.7                         0.427   
2937                11.0               11.2                         0.434   

      Schooling  
0          10.1  
1          10.0  
2           9.9  
3           9.8  
4           9.5  
...         ...  
2933        9.2  
2934        9.5  
2935       10.0  
2936        9.8  
2937        9.8  

[2938 rows x 22 columns]

In [8]:
data = data.fillna(data.mean())

# storing the cleaned data into a new variable (data)

In [9]:
data_no_mv = data

# storing that variable in a new variable

In [10]:
data_no_mv.columns

# getting the columns from the dataset listed

Index(['Country', 'Year', 'Status', 'LifeExpectancy', 'AdultMortality',
       'infantDeaths', 'Alcohol', 'PercentageExpenditure', 'HepatitisB',
       'Measles', 'BMI', 'UnderFiveDeaths', 'Polio', 'TotalExpenditure',
       'Diphtheria', ' HIV', 'GDP', 'Population', 'Thinness1to19Years',
       'Thinness5to9Years', 'IncomeCompositionOfResources', 'Schooling'],
      dtype='object')

## Standardising

In [11]:
x = data[['AdultMortality', 'infantDeaths', 'Alcohol', 'PercentageExpenditure', 'HepatitisB',
          'Measles', 'BMI', 'UnderFiveDeaths', 'Polio', 'TotalExpenditure', 'Diphtheria',
          ' HIV', 'GDP', 'Population', 'Thinness1to19Years', 'Thinness5to9Years',
          'IncomeCompositionOfResources', 'Schooling']]
y = data['LifeExpectancy']

# separating the independant variables (x) from the dependant variable (Life expectancy/y)

In [12]:
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

# importing a scaler from the sklearn library

In [13]:
scaler.fit(x)

# fitting the standard scaler to the dependant variables

StandardScaler()

In [14]:
x_scaled = scaler.transform(x)

# storing the newly scaled values into a new variable (x_scaled)

In [15]:
x_scaled

array([[ 0.79158632,  0.26882378, -1.1729584 , ...,  2.77327898,
        -0.72540055, -0.57993072],
       [ 0.85607167,  0.28578638, -1.1729584 , ...,  2.81790246,
        -0.74005007, -0.61056961],
       [ 0.83188966,  0.30274898, -1.1729584 , ...,  2.86252595,
        -0.7693491 , -0.6412085 ],
       ...,
       [-0.73994077, -0.04498439, -0.04414645, ..., -0.79659991,
        -0.97932554, -0.61056961],
       [ 4.20124926, -0.04498439, -0.73624609, ..., -0.70735294,
        -0.97932554, -0.6718474 ],
       [ 4.03197521, -0.05346569, -0.74646158, ...,  1.41226265,
        -0.94514333, -0.6718474 ]])

In [16]:
x_with_const = sm.add_constant(x_scaled)

In [17]:
# Fitting the model
model = sm.OLS(y, x_with_const).fit()

In [18]:
# Print summary statistics
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:         LifeExpectancy   R-squared:                       0.818
Model:                            OLS   Adj. R-squared:                  0.816
Method:                 Least Squares   F-statistic:                     726.7
Date:                Sat, 10 Feb 2024   Prob (F-statistic):               0.00
Time:                        11:22:33   Log-Likelihood:                -8285.8
No. Observations:                2938   AIC:                         1.661e+04
Df Residuals:                    2919   BIC:                         1.672e+04
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         69.2249      0.075    921.120      0.0

## Creating Multiple Linear Regression

In [19]:
reg = LinearRegression()
reg.fit(x_scaled,y)

# storing the regression model in the variable (reg) and fitting it to the scaled dependant variables and the independant one

LinearRegression()

In [20]:
reg.coef_

# the numbers below represent the coefficient of each independant variable
# this is the constant that is multiplied by the variable

array([-2.52282307e+00,  1.17358973e+01,  4.75221623e-01,  2.66493507e-01,
       -3.18865654e-01, -2.34626029e-01,  8.64689327e-01, -1.19308307e+01,
        6.71204415e-01,  2.13854927e-01,  9.37690747e-01, -2.39386775e+00,
        4.81128299e-01,  2.33348716e-03, -3.65759996e-01,  2.72018880e-03,
        1.22731901e+00,  2.18552383e+00])

In [21]:
reg.intercept_

# the intercept is the point where the function crosses the y axis (the predicted value for y when x is 0)

69.22493169398906

In [22]:
reg.score(x_scaled,y)

#.score() returns the R-Squared of a linear regression (simple or multiple)

# however, the Adjusted R-Squared is more appropriate for multiple linear regerssion
# as it adjusts for multiple variables

0.8175560042141601

In [23]:
x_scaled.shape

# 2938 is the number of observations, and 18 the number of predictors (features)

(2938, 18)

In [24]:
r2 = reg.score(x,y)

# variable storing the R-Squared value


n = x.shape[0]

# the number of observations


p = x.shape[1]

# the number of predictors (18)


adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)
adjusted_r2

# plugging in the variables into the R-Squared formula

C:\Users\JoeHM\Documents\PythonAnaconda\Lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


-185021633.65535915

In [25]:
# We can conclude that the Adjusted R Squared is significantly lower than the R Squared
# Therefore one or more of the predictors has little to no explanatory power
# To find out which predictors are the least important we use feature selection:

## Feature Selection

In [26]:
from sklearn.feature_selection import f_regression
f_regression(x_scaled,y)

# importing feature_selection module from sklearn library
# f_regression creates simple linear regressions of each feature
# applying the f_regression to the features (x) and target (y)

# first array shows F-statistics for each of the regressions
# Second array shows the corresponding P-values

(array([2.76404441e+03, 1.17962373e+02, 5.31781766e+02, 5.00991321e+02,
        1.27192318e+02, 7.47555807e+01, 1.33620149e+03, 1.52925260e+02,
        7.94861308e+02, 1.32741290e+02, 8.57392617e+02, 1.31687939e+03,
        6.67887756e+02, 1.13267385e+00, 8.42328652e+02, 8.17241347e+02,
        2.70508159e+03, 3.07202002e+03]),
 array([0.00000000e+000, 5.71450848e-027, 2.78146391e-108, 1.38444938e-102,
        6.57048596e-029, 8.61324481e-018, 1.94642951e-241, 2.79586247e-034,
        5.66704978e-155, 4.51844426e-030, 1.39163450e-165, 1.51847141e-238,
        7.22920497e-133, 2.87293240e-001, 4.82278222e-163, 8.62336187e-159,
        0.00000000e+000, 0.00000000e+000]))

In [27]:
p_values = f_regression(x_scaled,y)[1]
p_values

# storing the p-values from the regression in the variable 'p_values'

array([0.00000000e+000, 5.71450848e-027, 2.78146391e-108, 1.38444938e-102,
       6.57048596e-029, 8.61324481e-018, 1.94642951e-241, 2.79586247e-034,
       5.66704978e-155, 4.51844426e-030, 1.39163450e-165, 1.51847141e-238,
       7.22920497e-133, 2.87293240e-001, 4.82278222e-163, 8.62336187e-159,
       0.00000000e+000, 0.00000000e+000])

In [28]:
p_values.round(3)

# as the output is written in scientific notation, the p_values will be rounded to 3 digits

array([0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   ,
       0.   , 0.   , 0.   , 0.   , 0.287, 0.   , 0.   , 0.   , 0.   ])

In [29]:
# p_values above 0.05 or 5% are not significant, whereas below 0.05 are significant
# Therefore, it seems that most of these variables are significant except 'Population' (29%) so will drop that one

## Summary Table with P-values

In [30]:
reg_summary = pd.DataFrame(data = x.columns.values, columns=['Features'])
reg_summary

Features
0                 AdultMortality
1                   infantDeaths
2                        Alcohol
3          PercentageExpenditure
4                     HepatitisB
5                        Measles
6                            BMI
7                UnderFiveDeaths
8                          Polio
9               TotalExpenditure
10                    Diphtheria
11                           HIV
12                           GDP
13                    Population
14            Thinness1to19Years
15             Thinness5to9Years
16  IncomeCompositionOfResources
17                     Schooling

In [31]:
reg_summary ['Coefficients'] = reg.coef_
reg_summary ['p-values'] = p_values.round(3)

In [32]:
reg_summary

Features  Coefficients  p-values
0                 AdultMortality     -2.522823     0.000
1                   infantDeaths     11.735897     0.000
2                        Alcohol      0.475222     0.000
3          PercentageExpenditure      0.266494     0.000
4                     HepatitisB     -0.318866     0.000
5                        Measles     -0.234626     0.000
6                            BMI      0.864689     0.000
7                UnderFiveDeaths    -11.930831     0.000
8                          Polio      0.671204     0.000
9               TotalExpenditure      0.213855     0.000
10                    Diphtheria      0.937691     0.000
11                           HIV     -2.393868     0.000
12                           GDP      0.481128     0.000
13                    Population      0.002333     0.287
14            Thinness1to19Years     -0.365760     0.000
15             Thinness5to9Years      0.002720     0.000
16  IncomeCompositionOfResources      1.227319     0.000
17                     Schooling      2.185524     0.000

### Conclusion:

##### Significance of Features:

##### The p-values associated with each coefficient are all very low (0.000), indicating that all features are statistically significant in predicting life expectancy. This means that there is strong evidence to suggest that these features have a significant impact on life expectancy.
##### Impact of Features:

##### Positive coefficients indicate that an increase in the corresponding feature is associated with an increase in life expectancy. For example, higher values of features like infant deaths, alcohol consumption, percentage expenditure on healthcare, BMI, polio vaccination coverage, GDP, etc., are associated with higher life expectancy.
##### Negative coefficients suggest that an increase in the corresponding feature is associated with a decrease in life expectancy. For instance, higher values of features like adult mortality rate, measles incidence, under-five deaths, HIV prevalence, thinness in 1-19 years, etc., are associated with lower life expectancy.
##### Relative Importance:

##### The magnitude of the coefficients provides insights into the relative importance of each feature in predicting life expectancy. Features with larger coefficients have a stronger impact on life expectancy compared to those with smaller coefficients.
##### Non-Significant Feature:

##### The feature "Population" has a p-value of 0.287, which is higher than the typical significance level (e.g., 0.05). This suggests that population may not be a significant predictor of life expectancy in this model, as its coefficient is not statistically different from zero at the 0.05 significance level.
##### Interpretation:

##### Based on the coefficients, one could interpret the impact of each feature on life expectancy. For example, a one-unit increase in schooling is associated with a 2.18 year increase in life expectancy, holding all other variables constant.
##### In summary, the regression analysis suggests that various factors such as healthcare indicators, socioeconomic factors, disease prevalence, and nutrition have significant impacts on life expectancy. Policymakers and public health officials can use this information to prioritize interventions and policies aimed at improving these determinants of population health and ultimately increasing life expectancy.